### ISSUE TITLE
[java] UnusedPrivateMethod - false negative with anonymous classes

### ISSUE THREAD

**Issue by Jane-PL-256** [pmd/pmd#3627]
**Affects PMD Version:6.39.0** 

**Rule:UnusedPrivateMethod**

Please provide the rule name and a link to the rule documentation:
<https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedprivatemethod>

**Description:**
I think PMD should report a warning at line 3 because this is a unused private method, but it doesn't. Thanks for your consideration. I am happy to provide my help.

**Code Sample demonstrating the issue:**
```java
public class Tester {
  Object anonymousClass = new Object() {
      private void foo() {}  // unused, should report a warning
  };
}
```

**Expected outcome: A warning**

PMD should report a violation at line 3, but doesn't. This is a false-negative.

**Running PMD through:** *[CLI]*

**Comment by Belle-PL** [comment]
@adangel Hi Andreas! Sorry to bother you. I think this issue and #3626 is similar to #3618. Could you please talk about you feelings or provide help? 😄 If possible, I also want to try to fix these issues and propose a PR. Thank you.

**Comment by adangel** [comment]
Hi @Belle-PL , thanks for your offer. Yes, these cases are similar. However, the solution might be different. E.g. the problem with UnusedFormalParameter was (fixed in #3619) that it didn't traverse the whole tree (with `super.visit(...)`) but skipped it simply (with `return data`). That's why methods declared inside methods (such as in anonymous classes) were not visited.

UnusedPrivateMethodRule uses the rulechain for "ASTClassOrInterfaceDeclaration", which means, all nodes of this type are visited and you don't need to call `super.visit(...)`. However, anonymous classes don't have a ClassOrInterfaceDeclaration, they just have a ClassOrInterfaceBody. So maybe this node should be visited instead (but then you need to adjust e.g. the way how to figure out whether this is a interface or not - but: private default methods in interface are probably currently also not detected but should be, so maybe it doesn't matter whether this is interface or not...)

UnusedPrivateFieldRule seems to suffer from the same problem - only visiting ClassOrInterfaceDeclaration which doesn't exist for anonymous classes.

**Comment by jsotuyod** [comment]
This is no longer reproducible under PMD 7.0.0
